### PR TITLE
bpo-28754: AC on bisect.

### PR DESCRIPTION
**Pull request by JulienPalard**
[[bpo-28754](https://bugs.python.org/issue28754)]

### CODE REVIEW

**Review by methane — COMMENTED** [review]
LGTM.
But please wait approval from expert.
(@rhettinger is listed on http://cpython-devguide.readthedocs.io/experts.html)

File: Modules/_bisectmodule.c [hunk 1]
```diff
@@ -6,6 +6,39 @@ Converted to C by Dmitry Vasiliev (dima at hlabs.spb.ru).
 #define PY_SSIZE_T_CLEAN
 #include "Python.h"
 
+static int
+ssize_t_converter(PyObject *obj, void *ptr)
```
Comment (serhiy-storchaka) [inline Modules/_bisectmodule.c:10]: ssize_t_converter is not used for parsing arguments. It is used only in optional_ssize_t_converter. Wouldn't be better to inline it?

File: Modules/_bisectmodule.c [hunk 2]
```diff
@@ -44,56 +77,79 @@ internal_bisect_right(PyObject *list, PyObject *item, Py_ssize_t lo, Py_ssize_t
...
+[python start generated code]*/
+/*[python end generated code: output=da39a3ee5e6b4b0d input=53c711121eb30d3f]*/
+
+
+/*[clinic input]
+bisect.bisect_right
+
+    a: object
+    x: object
+    lo: Py_ssize_t(c_default='0') = 0
```
Comment (serhiy-storchaka) [inline Modules/_bisectmodule.c:94]: `c_default='0'` is redundant.

File: Modules/_bisectmodule.c [hunk 3]
```diff
@@ -44,56 +77,79 @@ internal_bisect_right(PyObject *list, PyObject *item, Py_ssize_t lo, Py_ssize_t
...
+class hi_parameter_converter(CConverter):
+    type = 'Py_ssize_t'
+    converter = 'optional_ssize_t_converter'
+
+[python start generated code]*/
+/*[python end generated code: output=da39a3ee5e6b4b0d input=53c711121eb30d3f]*/
+
+
+/*[clinic input]
+bisect.bisect_right
```
Comment (serhiy-storchaka) [inline Modules/_bisectmodule.c:90]: You can use the return converter for simplifying the code.
```
bisect.bisect_right -> Py_ssize_t
```

The body of `bisect_bisect_right_impl()` would be just:
```
return internal_bisect_right(a, x, lo, hi);
```

File: Lib/test/test_bisect.py [hunk 4]
```diff
@@ -186,6 +186,11 @@ def test_optionalSlicing(self):
                         self.assertTrue(data[ip-1] <= elem)
                     self.assertEqual(ip, max(lo, min(hi, expected)))
 
+    def test_hi_backcompatbility(self):
+        for func, data, elem, expected in self.precomputedCases:
+            self.assertEqual(func(data, elem, 0, -1),
```
Comment (serhiy-storchaka) [inline Lib/test/test_bisect.py:191]: This doesn't check that passing `-1` and `None` is the same as not specifying the hi argument. Compare results with `func(data, elem)`.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

**Comment by rhettinger** [comment]
I don't want the pure python API to change at all.  It is a long standing API and there is nothing wrong with it.  Also, passing it -1 here as a special value makes real code less readable.  Further, the change slightly slows down the case where high is specified.  In a high performance building block module, we shouldn't add in unnecessary checks.

**Comment by brettcannon** [comment]
To try and help move older pull requests forward, we are going through and backfilling 'awaiting' labels on pull requests that are lacking the label. Based on the current reviews, the best we can tell in an automated fashion is that a core developer requested changes to be made to this pull request.

If/when the requested changes have been made, please leave a comment that says, `I have made the requested changes; please review again`. That will trigger a bot to flag this pull request as ready for a follow-up review.

**Pull request closed without merging**